sqlx-sqlite/src/connection/executor.rs

```diff
@@ -64,13 +64,15 @@ impl<'c> Executor<'c> for &'c mut SqliteConnection {
 
             futures_util::pin_mut!(stream);
 
+            let mut out = Ok(None);
+
             while let Some(res) = stream.try_next().await? {
                 if let Either::Right(row) = res {
-                    return Ok(Some(row));
+                    out = Ok(Some(row));
```

Comment:
The reason it doesn't work this way is because we didn't want the application to have to wait for the full result set before returning. For example, if someone forgot a `LIMIT 1` on their query and was doing a `SELECT ... ORDER BY ...` to get the first/last item in an ordered set, the call would have to churn through the whole dataset before returning.

The bug happens because of a race condition: the worker sends the row data and then continues calling `sqlite3_step()` and only stops when it notices the result channel is closed, then it calls `sqlite3_reset()` which is supposed to actually commit the changes. In the meantime, the application receives the row and then continues to the next query, which it executes on another connection which may not be able to see the changes yet.

In reality, this needs to be handled proactively by the driver. The message to the worker should specify that it's only expecting one row, and then the worker should step until it gets the row and immediately call `sqlite3_reset()` (by dropping `ExecuteIter`) before returning it. That should fix the race condition.

Reply:
That makes sense, thanks @abonander. Does the approach in [02ca0ed](https://github.com/launchbadge/sqlx/pull/3354/commits/02ca0edd358047828a8d5a04a16ab714ed6b4829) look a little bit closer to what you had in mind? I don't know the project's naming/organization conventions so apologies for that.